text/1015-full-text-search.rst

```diff
@@ -73,59 +84,192 @@ queries rather than introducing a new scalar type. This string is expected to
+      index fts::textsearch(language := 'english')
+        on (
+          fts::weight(.title, 'A'),
+          fts::weight(.body, 'B')
+        );
+  }
 
-  function fts::match(
-    doc: Object,
-    query: str,
-    rank_opts: str = 'default',
-    highlight_opts: optional str = {}
-  ) -> FTSResult
+The ``fts::weight`` pseudo-function cannot be called normally, but in index
+specification it associates a weight group with a particular indexed
+expression (typically just a property). Using strings to denote weight groups
+makes it less likely that the users confuse the *group name* with the actual
+associated weight *value*.
+
+The most basic naming scheme for weight groups could be similar to Postgres:
+just using the alphabet starting with 'A'. We can have more than 4 groups,
+though for the general case. If the backend supports fewer groups than
+indicated by index specification, we simply use as many groups as we can
+starting with 'A' and merge the rest into the last valid group ('D' for
+Postgres).
+
+In principle the group names can be generalized to be any ``str`` and the
+lexicographical order can then be used to determine the relative group order
+(for purpose of picking group weights). This would allow using Postgres-style
+groups 'A', 'B', 'C', 'D' or give groups more descriptive names like '1 - most
+relevant', '2 - regular', '3 - marginal'.
+
+
+Inheritance
+-----------
+
+Since ``fts::textsearch`` index is semantic and has effect on how FTS
+functions work, we cannot have multiple versions of this index defined on a
+single object type. So at most one FTS index can be defined on any type
+(subject to ``??`` once the extendion enabling/disabling is implemented). In
+particular this means that when an FTS index is inherited by a type and a new
+index definition is provided in the extending type, the new definition
+completely overrides the inherited one.
+
+For example::
+
+  type Document {
+      body: str;
+      index fts::textsearch(language := 'english') on (.body);
+  }
+
+  type InternalDoc extending Document {
+      internal: str;
+  }
+
+  type TitledDoc extending Document {
+      title: str;
+      index fts::textsearch(language := 'english') on (.title);
+  }
+
+The ``InternalDoc`` type has no FTS index of its own, so it inherits the index
+from ``Document``. Thus only the ``.body`` is indexed. Conversely,
+``TitledDoc`` defines a new ``fts::textsearch`` index on ``.title``, but the
+``body`` property would no longer be indexed for ``TitledDoc`` objects.
+
+
+Search Functions
+----------------
+
+An important feature of the search function is that it actually performs the
+job typically done by the ``FILTER`` clause. The function is supposed to
+filter a bunch of objects based on the FTS query and indexed fields. In
+addition to filtering, the results may be annotated with relevance (ranked)
+and potentially with highlighted matches::
+
+  function std::fts::search(
+      docs: set of anyobject,
+      query: str,
```

Comment:
I believe we decided we needed a `language` param also

Reply:
Added parameter